src/main/kotlin/Dna.kt

```diff
@@ -0,0 +1,24 @@
+class Dna(private val sequence: String) {
+
+    init {
+        if (!isValid(sequence)) {
+            throw IllegalArgumentException("Invalid DNA sequence: $sequence")
+        }
+    }
+
+    // Count occurrences of each valid nucleotide
+    val nucleotideCounts: Map<Char, Int>
```

Comment:
try refactor like:
` dna.groupingBy {  }.eachCountTo(mutableMapOf('A' to 0, 'C' to 0, 'G' to 0, 'T' to 0))`